RecoTauTag/HLTProducers/src/L2TauTagNNProducer.cc

```diff
@@ -678,30 +662,18 @@ void L2TauNNProducer::fillPatatracks(tensorflow::Tensor& cellGridMatrix,
   auto getCell = [&](NNInputs input) -> float& {
     return getCellImpl(cellGridMatrix, tau_idx, phi_idx, eta_idx, input);
   };
+
+  std::vector<int> TrkGood;
+  std::vector<int> VtxGood;
+
+  selectGoodTracksAndVertices(patavtx_soa, patatracks_tsoa, TrkGood, VtxGood);
+
   const int nTaus = static_cast<int>(allTaus.size());
   for (tau_idx = 0; tau_idx < nTaus; tau_idx++) {
     const float tauEta = allTaus[tau_idx]->eta();
     const float tauPhi = allTaus[tau_idx]->phi();
 
-    auto maxTracks = patatracks_tsoa.stride();
-    auto const* quality = patatracks_tsoa.qualityData();
-
-    std::vector<int> TrackGood;
-    for (int32_t it = 0; it < maxTracks; ++it) {
-      auto nHits = patatracks_tsoa.nHits(it);
-      if (nHits == 0)
-        break;
-      auto q = quality[it];
-      if (q < pixelTrack::Quality::loose)
-        continue;
-      if (nHits < 0)
-        continue;
-      TrackGood.push_back(it);
-    }
-
-    std::vector<int> VtxGood = selectGoodVertices(patavtx_soa, patatracks_tsoa, TrackGood);
-
-    for (const auto it : TrackGood) {
+    for (const auto it : TrkGood) {
```

Comment:
This is now VERY different the the previous version.
Now only tracks contributing to the vertex fidning in patatrack (i.e not-small-pt HighPurity quadruplets( are considered.

Before all  tracks (including loose triplets) were used.
So I do not see how this code can reproduce previous results even on CPU

Reply:
@VinInn sorry, can you, please, elaborate on what is the difference? In the [master](https://github.com/cms-sw/cmssw/blob/master/RecoTauTag/HLTProducers/src/L2TauTagNNProducer.cc#L690-L700) `TrackGood ` are tracks with `quality >= loose && nHits > 0`. In the current version, `TrkGood` have exactly the same definition as far as I see...https://github.com/cms-sw/cmssw/blob/d8d790c8805a7e2d62a8b517b43664ca02c61a35/RecoTauTag/HLTProducers/src/L2TauTagNNProducer.cc#L598-L603

Reply:
Sorry I missread the position of the closing bracket of line 588
```
if (vtx_ass_to_track >= 0 && vtx_ass_to_track <= nv) {
```

so ok.
Still I would invita you to reconsider the use of all tracks (including loose triplets far away from any vertex) in your CNN
(and eventually clamp the pt of the track there as well).

Please also note that after the introduction of #36215 the  correspondence of nLayers with nHits has changed as tracks now can have 2 hits in each layer. So you may wish to use nLayers now in the CNN.

Reply:
@VinInn  thank you for the suggestion! For the next iteration of the training, we'll consider applying a cleaning of the input tracks. Since the current training was done using all tracks that pass `quality >= loose && nHits > 0`, we prefer to use the same selection for the inference. If I interpret correctly the last @valeriadamante results, the presence of such low-quality tracks does not represent an issue as far as reproducibility is concerned:i.e. small differences in inputs result in small differences in the NN score. So the original problem was due to the missing `break` for `nHits == 0`.

About `nHits` vs. `nLayers`: apart from preselection (which shouldn't be sensitive to this since we compare to 0), we use nHits in the definition of track ndof: `patatrackNdof = 2 * nHits - 5`. Could you, please, confirm that we need to switch to `nLayers` in this formula to get the correct ndof of the track?

Reply:
The correct one should be
```patatrackNdof = 2 * std::min(6,nHits) - 5```  as we fit 6 hits max.
BUT I should check that this is consistently done in patatrack itself...

As long as you did not give nHIts as direct input to the CNN no other change is needed (the CNN should not be very sensitive to how many hits we fit)